### PR TITLE
BD rhapsody sequence analysis

### PR DESCRIPTION
**Pull request by rcannood**
## Issue ticket number

Closes #xxxx 

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

- [ ] Conforms to the [Contributing guidelines](https://github.com/viash-hub/base/blob/main/CONTRIBUTING.md)

- [ ] Proposed changes are described in the [CHANGELOG.md](https://github.com/viash-hub/base/blob/main/CHANGELOG.md)

- [ ] I have tested my code with `viash ns test --parallel -q <name or namespace>`

- Check the correct box. Does this PR contain:
  - [ ] Breaking changes
  - [ ] New functionality
  - [ ] Major changes
  - [ ] Minor changes
  - [ ] Documentation
  - [ ] Bug fixes

### CODE REVIEW

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        if meta["cpus"]:
+            cwl_data = re.sub('"coresMin": [^\n]*[^,](,?)\n', f'"coresMin": {meta["cpus"]}\\1\n', cwl_data)
+
+        # Write the file out again
+        with open(cwl_file, 'w') as file:
+            file.write(cwl_data)
+    else:
+        cwl_file = orig_cwl_file
+
+    return os.path.abspath(cwl_file)
```
Comment (DriesSchaumont) [inline src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:159]: I prefer to return early 🤷 
```suggestion
    orig_cwl_file=os.path.join(meta["resources_dir"], "rhapsody_pipeline_2.2.1_nodocker.cwl")
    if not meta["memory_mb"] and not meta["cpus"]:
        return os.path.abspath(orig_cwl_file)
     
    # Inject computational requirements into pipeline
    cwl_file = os.path.join(dir, "pipeline.cwl")

    # Read in the file
    with open(orig_cwl_file, 'r') as file :
        cwl_data = file.read()

    # Inject computational requirements into pipeline
    if meta["memory_mb"]:
        memory = int(meta["memory_mb"]) - 2000 # keep 2gb for OS
        cwl_data = re.sub('"ramMin": [^\n]*[^,](,?)\n', f'"ramMin": {memory}\\1\n', cwl_data)
    if meta["cpus"]:
        cwl_data = re.sub('"coresMin": [^\n]*[^,](,?)\n', f'"coresMin": {meta["cpus"]}\\1\n', cwl_data)

    # Write the file out again
    with open(cwl_file, 'w') as file:
        file.write(cwl_data)
        
    return os.path.abspath(cwl_file)
```

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    } else {
+      out <- map(x, remove_docker)
+      out2 <- out[!sapply(out, is.null)]
+      return(out2)
+    }
+  } else {
+    return(x)
+  }
+}
+cwl_obj_2 <- remove_docker(cwl_obj)
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R:22]: you can indeed run `cwl-runner --no-container ...` but it used to be the case that the different steps had the Docker container marked as required and so the pipeline would fail anyway. Hang on, let me try again.

[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R:22]: Seems to be fixed now! I see that the DockerRequirements have been moved to hints as opposed to strict requirements :relaxed:

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R:1]: 🙅 No thank you! It will never be used as part of a pipeline or a standalone component. The script uses just base tidyverse. And oh, one dynutils statement which probably has an equivalent function in the tidyverse.

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+def copy_outputs(par: dict[str, Any], config: dict[str, Any]):
+    for arg in config["arguments"]:
+        par_value = par[arg["clean_name"]]
+        if par_value and arg["type"] == "file" and arg["direction"] == "output":
+            # example template: '[sample_name]_(assay)_cell_type_experimental.csv'
+            template = (arg.get("info") or {}).get("template")
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:166]: It's a pitfall but this doesn't work! Because `info` might be `null` which causes `arg.get("info", {})` to return `None`.

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        |cwl:tool: rhapsody
+        |""")]
+
+    for argument in config["arguments"]:
+        arg_info = argument.get("info") or {}
+        config_key = arg_info.get("config_key")
+        if par[argument["clean_name"]] and config_key:
+
+            if argument["type"] == "file":
+                str = strip_margin(f"""\
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py]: :+1:

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+def generate_config(par: dict[str, Any], config) -> str:
+    content_list = [strip_margin(f"""\
+        |#!/usr/bin/env cwl-runner
+        |
+        |cwl:tool: rhapsody
+        |""")]
+
+    for argument in config["arguments"]:
+        arg_info = argument.get("info") or {}
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:105]: see below :relaxed:

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    env = dict(os.environ)
+    env["TMPDIR"] = temp_dir
+
+    # Create output dir if not exists
+    if not os.path.exists(par["output_dir"]):
+        os.makedirs(par["output_dir"])
+
+    # Run command
+    print("> " + ' '.join(cmd), flush=True)
+    _ = subprocess.check_call(
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:231]: Do you mean `out = subprocess.check_output(cmd_pars).decode("utf-8")` and then printing the stdout?

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R]: Wohoo! that should clean things up 👍

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:173]: Ah, so if I understand correctly you get a TypeError when calling `.get("template")`.  Could you add a small comment

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:231]: Yes, printing it, or writing it to an output file.

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/config.vsh.yaml [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+namespace: bd_rhapsody
+description: |
+  BD Rhapsody Sequence Analysis CWL pipeline v2.2.
+
+  This pipeline performs analysis of single-cell multiomic sequence read (FASTQ) data. The supported
+  sequencing libraries are those generated by the BD Rhapsody™ assay kits, including: Whole Transcriptome
+  mRNA (WTA), Targeted mRNA, AbSeq Antibody-Oligonucleotides (ABC), Single-Cell Multiplexing (SMK),
+  TCR/BCR (VDJ), and ATAC-Seq.
+
+  The CWL pipeline file is obtained by cloning 'https://bitbucket.org/CRSwDev/cwl' and removing all objects with class 'DockerRequirement' from the YAML.
```
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/config.vsh.yaml:11]: good point!

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/_process_cwl.R:1]: done!

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:231]: I want to view the output as it comes out of the process ^^

[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:231]: Printing it is also fine. But right now, neither one of those is happening (because check_call is used). https://docs.python.org/3/library/subprocess.html#subprocess.check_call

```
Code needing to capture stdout or stderr should use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) instead:
```

**Review by DriesSchaumont — CHANGES_REQUESTED** [review]

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py:231]: I don't understand. `subprocess.check_call` does exactly what I want it to do.

```python
>>> out = subprocess.check_call(["echo", "hi"])
hi
```

For sure, `subprocess.run` seems to do exactly the same:

```python
>>> out = subprocess.run(["echo", "hi"], check=True)
hi
```

I guess I don't understand what the difference between the two variants is and why you'd like me to use `subprocess.run` over `subprocess.check_call`.

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py]: Added note. Sorry for the delay!

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rcannood) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py]: Added note! :relaxed:

File: src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DriesSchaumont) [reply src/bd_rhapsody/bd_rhapsody_sequence_analysis/script.py]: Turns out that the pipeline already returns its own log in the output, so this has become a non-issue 👍

**Review by DriesSchaumont — APPROVED** [review]
